Add platform test for kdump

amnsinghal opened the pull request:
Summary:
Add test_kdump.py to platform_tests.

### Type of change
- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [X] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [X] 202205

### Approach
#### What is the motivation for this PR?
Add automation test for kdump functionality.

#### How did you do it?
Induce kernel panic thru sysrq-trigger.

#### How did you verify/test it?
For platform with kdump enabled, trigger kernel panic, and check for reboot-cause on subsequent bootup.
Validated on Cisco-8000 platforms thru internal sonic-mgmt test.

![image](https://github.com/sonic-net/sonic-mgmt/assets/127834195/0bfe9b34-af29-4321-bcc6-923fee2f2e71)

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
Any platform with kdump enabled, else test will be skipped.

amnsinghal commented:
@prgeor @lguohan @abdosi Pls help review this new testcase for kdump, thanks.

amnsinghal commented:
@prgeor Can you pls help check this Elastictest failure, doesn't seem related to my change, thanks.

amnsinghal commented:
@prgeor Can you pls help review and check this Elastictest failure, doesn't seem related to my change, thanks.

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/platform_tests/test_kdump.py:21:1: E302 expected 2 blank lines, found 1<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

bpar9 commented on tests/platform_tests/test_kdump.py:
```diff
@@ -0,0 +1,117 @@
+        dut_datetime = duthost.get_now_time()
+
+        out = duthost.command('show kdump config')
+        if "Enabled" not in out["stdout"]:
```
do we support the entire behavior in this document - https://github.com/sonic-net/SONiC/blob/master/doc/kdump/SONiC-kdump.md . If so, then why skip the test if disabled, cant we enable kdump with CLI and then proceed with the test -
root@croc-xx36-dut:/home/cisco# config kdump ?
Usage: config kdump [OPTIONS] COMMAND [ARGS]...

Configure the KDUMP mechanism

Options:
-h, -?, --help Show this message and exit.

Commands:
disable Disable the KDUMP mechanism
enable Enable the KDUMP mechanism
memory Configure the memory for KDUMP mechanism
num_dumps Configure the maximum dump files of KDUMP mechanism

amnsinghal replied:
This test will run only if kdump is already enabled, as we are not sure if other platforms support it.

bpar9 commented on tests/platform_tests/test_kdump.py:
```diff
@@ -0,0 +1,117 @@
+        pytest_assert(dut_uptime > dut_datetime, "Device {} did not reboot".format(hostname))
+
+        out = duthost.command('show reboot-cause')
+        if "Kernel Panic" not in out["stdout"]:
```
can you use the common proc - check_reboot_cause from common/reboot.py

amnsinghal replied:
Should I add 'Kernel Panic' to ```reboot_ctrl_dict``` ?

amnsinghal replied:
Updated.

bpar9 commented:
@yxieca there is an old PR which has a superset of TCs then what is present in this PR. That PR has been reviewed but not merged for past couple of years- https://github.com/sonic-net/sonic-mgmt/pull/3562 . Is there some reason why community has not merged the old PR

yxieca commented:
> @yxieca there is an old PR which has a superset of TCs then what is present in this PR. That PR has been reviewed but not merged for past couple of years- #3562 . Is there some reason why community has not merged the old PR

Good question. Thanks for looking up the history. I believe the old PR #3562 has unresolve review comments. That was the reason for leaving it open.

bpar9 commented:
> > @yxieca there is an old PR which has a superset of TCs then what is present in this PR. That PR has been reviewed but not merged for past couple of years- #3562 . Is there some reason why community has not merged the old PR
> 
> Good question. Thanks for looking up the history. I believe the old PR #3562 has unresolve review comments. That was the reason for leaving it open.

thanks @yxieca can the reply. So what is your guidance on this PR-10047? it has a subset of the TCs of #3562 which are currently applicable for our platform. So do we wait for 3562 to be merged or can we go ahead and push this PR in. Please let us know, so that we can proceed accordingly. Thanks

yxieca commented:
> > > @yxieca there is an old PR which has a superset of TCs then what is present in this PR. That PR has been reviewed but not merged for past couple of years- #3562 . Is there some reason why community has not merged the old PR
> > 
> > 
> > Good question. Thanks for looking up the history. I believe the old PR #3562 has unresolve review comments. That was the reason for leaving it open.
> 
> thanks @yxieca for the reply. So what is your guidance on this PR-10047? it has a subset of the TCs of #3562 which are currently applicable for our platform. So do we wait for 3562 to be merged or can we go ahead and push this PR in. Please let us know, so that we can proceed accordingly. Thanks

We can work on this PR and get it merged. At this point. this PR contains blocking comments. Please resolve.

bpar9 commented:
> > > > @yxieca there is an old PR which has a superset of TCs then what is present in this PR. That PR has been reviewed but not merged for past couple of years- #3562 . Is there some reason why community has not merged the old PR
> > > 
> > > 
> > > Good question. Thanks for looking up the history. I believe the old PR #3562 has unresolve review comments. That was the reason for leaving it open.
> > 
> > 
> > thanks @yxieca for the reply. So what is your guidance on this PR-10047? it has a subset of the TCs of #3562 which are currently applicable for our platform. So do we wait for 3562 to be merged or can we go ahead and push this PR in. Please let us know, so that we can proceed accordingly. Thanks
> 
> We can work on this PR and get it merged. At this point. this PR contains blocking comments. Please resolve.

thanks @yxieca , Aman will work on addressing  the comments and proceed.

amnsinghal commented:
> > > > > @yxieca there is an old PR which has a superset of TCs then what is present in this PR. That PR has been reviewed but not merged for past couple of years- #3562 . Is there some reason why community has not merged the old PR
> > > > 
> > > > 
> > > > Good question. Thanks for looking up the history. I believe the old PR #3562 has unresolve review comments. That was the reason for leaving it open.
> > > 
> > > 
> > > thanks @yxieca for the reply. So what is your guidance on this PR-10047? it has a subset of the TCs of #3562 which are currently applicable for our platform. So do we wait for 3562 to be merged or can we go ahead and push this PR in. Please let us know, so that we can proceed accordingly. Thanks
> > 
> > 
> > We can work on this PR and get it merged. At this point. this PR contains blocking comments. Please resolve.
> 
> thanks @yxieca , Aman will work on addressing the comments and proceed.

Thanks for confirming @yxieca and @bpar9, I have resolved the comments and update the PR, pls review and approve/merge.

amnsinghal commented:
@yxieca All checks have passed, except Semgrep is still waiting for status to be reported, can you pls check this. Thanks!

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/platform_tests/test_kdump.py:61:9: F841 local variable 'dut_ip' is assigned to but never used<br/>tests/platform_tests/test_kdump.py:63:9: F841 local variable 'dut_datetime' is assigned to but never used<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

yxieca commented:
@prgeor can you review this test?

amnsinghal commented:
@prgeor Can you pls review and approve/merge, if no further suggestions, thanks!

prgeor commented on tests/platform_tests/test_kdump.py:
```diff
@@ -0,0 +1,102 @@
+    """
+    This test case is used to verify that DUT will load kdump crashkernel on kernel panic.
+    """
+    def wait_lc_healthy_if_sup(self, duthost, duthosts, localhost, conn_graph_facts, xcvr_skip_list):
```
@amnsinghal is this test supposed to run only on supervisor?

bmridul replied:
No, the test runs on both LCs and RPs. For RPs, there is an additional check to ensure that LCs come up healthy after RP goes down and come back up. (In SONiC distributed chassis, LCs go down as well if RP goes down - Ref: https://github.com/sonic-net/SONiC/pull/945#discussion_r828304573)

prgeor commented on tests/platform_tests/test_kdump.py:
```diff
@@ -0,0 +1,102 @@
+        if "Enabled" not in out["stdout"]:
+            pytest.skip('DUT {}: Skip test since kdump is not enabled'.format(hostname))
+
+        reboot(duthost, localhost, reboot_type=REBOOT_TYPE_KERNEL_PANIC)
```
@amnsinghal why not trigger panic via `echo c > /proc/sysrq-trigger`

bmridul replied:
That is indeed what is being done with "REBOOT_TYPE_KERNEL_PANIC". See changes in reboot.py above.

amulyan7 commented:
@prgeor if no further comments, can you please approve/merge, thanks!

bpar9 commented:
@prgeor , @yxieca , as we addressed all the comments, can we go ahead and merge this? thanks

shivuv commented:
> @prgeor , @yxieca , as we addressed all the comments, can we go ahead and merge this? thanks

Team: Can you please merge this PR ?

shivuv commented:
@prgeor  @yxieca  @gechiang @abdosi : Please help to merge this PR into master and cherry pick this into 202305 and 202311 branches.

abdosi commented:
@wangxin please help with cherry-pick into 202205/202305.

mssonicbld commented:
Cherry-pick PR to 202205: https://github.com/sonic-net/sonic-mgmt/pull/11485

mssonicbld commented:
Cherry-pick PR to 202305: https://github.com/sonic-net/sonic-mgmt/pull/11486
